### PR TITLE
1D vectors instead of nD in Fully-Connected layer

### LINKED ISSUE #38: Use 1D vector instead of nD for matrices

### LINKED ISSUE #39: Implement Shape data structure class

### PR DESCRIPTION
**Pull request by NeiroYT**
Closes #38 
Closes #39

### CODE REVIEW

File: include/layers/FCLayer.hpp [hunk 1]
```diff
@@ -2,27 +2,78 @@
 #include <stdexcept>
 #include <vector>
 
+template <typename T>
+class Shape2D {
```
Comment (aobolensk) [inline include/layers/FCLayer.hpp:6]: Shapes can be not only 2D. Can we consider that?

File: include/layers/FCLayer.hpp [hunk 2]
```diff
@@ -111,17 +139,17 @@ FCLayer<ValueType>& FCLayer<ValueType>::operator=(const FCLayer& sec) {
...
   }
-  std::vector<ValueType> output_values = mat_vec_mul(weights_, input);
+  Shape2D<ValueType> output_values = mat_vec_mul(weights_, input);
   for (size_t i = 0; i < this->outputSize_; i++) {
-    output_values[i] += bias_[i];
+    output_values.set(i, 0, output_values.get(i, 0) + bias_.get(i, 0));
   }
   return output_values;
-}
+}
```
Comment (aobolensk) [inline include/layers/FCLayer.hpp:155]:
```suggestion
}

```

File: include/layers/FCLayer.hpp [hunk 3]
```diff
@@ -2,27 +2,78 @@
...
+  size_t get_height() const { return height_; }
+  size_t size() const { return width_ * height_; }
+  void resize(size_t height, size_t width) {
+    this->width_ = width;
+    this->height_ = height;
+    data_.resize(width_ * height_, T(0));
+  }
+
+ private:
+  std::vector<T> data_;
```
Comment (aobolensk) [inline include/layers/FCLayer.hpp:53]: Shapes are usually not connected to the data. They just define dimensions.

**Review by aobolensk — CHANGES_REQUESTED** [review]

File: include/layers/FCLayer.hpp (hunk 3 above)
Comment (NeiroYT) [reply include/layers/FCLayer.hpp:53]: If they aren't connected, abstract Layer's run() function will now return std::pair(vector<ValueType> data, Shape shape) ?

Comment (aobolensk) [reply include/layers/FCLayer.hpp:53]: Shape is actually a part of the layer description

Comment (aobolensk) [reply include/layers/FCLayer.hpp:53]: Meaning that we have input shape, output shape

File: include/layers/FCLayer.hpp (hunk 2 above)
Comment (NeiroYT) [reply include/layers/FCLayer.hpp]: Fixed

File: include/layers/FCLayer.hpp (hunk 1 above)
Comment (NeiroYT) [reply include/layers/FCLayer.hpp]: Now nD Shape is possible

File: include/layers/FCLayer.hpp (hunk 3 above)
Comment (NeiroYT) [reply include/layers/FCLayer.hpp]: Changed

File: src/layers/FCLayer.cpp [hunk 4]
```diff
@@ -1 +1,19 @@
 #include "layers/FCLayer.hpp"
+
+size_t Shape::get_index(const std::vector<size_t>& coords) const {
+  size_t res = 0;
+  size_t mulbuf;
```
Comment (aobolensk) [inline src/layers/FCLayer.cpp:5]: Why it needs to be defined here?

File: include/layers/FCLayer.hpp [hunk 5]
```diff
@@ -1,17 +1,62 @@
...
+  Shape(size_t dims_count) : dims_(dims_count, 0) {}
+  Shape(const std::vector<size_t>& dims) : dims_(dims) {}
+  Shape(const std::initializer_list<size_t>& l) : dims_(l) {}
+  Shape(const Shape& c) = default;
+  Shape& operator=(const Shape& c) = default;
+  size_t operator[](size_t i) const { return dims_[i]; }
+  size_t& operator[](size_t i) { return dims_[i]; }
+  size_t at(size_t i) const {
+    if (i >= dims_.size()) {
+      throw std::out_of_range("Bad shape index");
```
Comment (aobolensk) [inline include/layers/FCLayer.hpp:20]:
```suggestion
      throw std::out_of_range("Invalid shape index");
```
Let it be aligned, same for other places

**Review by aobolensk — CHANGES_REQUESTED** [review]

File: src/layers/FCLayer.cpp (hunk 4 above)
Comment (NeiroYT) [reply src/layers/FCLayer.cpp]: To separate complex non-template function from the header file

Comment (aobolensk) [reply src/layers/FCLayer.cpp]: I meant 'mulbuf' outside of for loop

Comment (NeiroYT) [reply src/layers/FCLayer.cpp]: Can also place it somewhere under the class definition

Comment (aobolensk) [reply src/layers/FCLayer.cpp]: It's about mulbuf variable

Comment (NeiroYT) [reply src/layers/FCLayer.cpp]: It's moved now

File: include/layers/FCLayer.hpp [hunk 6]
```diff
@@ -1,17 +1,64 @@
 #pragma once
+#include <algorithm>
+#include <initializer_list>
+#include <numeric>
 #include <stdexcept>
 #include <vector>
 
+// height -> width -> other dims
```
Comment (aobolensk) [inline include/layers/FCLayer.hpp:8]: Not necessary this layout. Height and width could be not the first two dims, it's better to remove this misleading comment
```suggestion
```

**Review by aobolensk — APPROVED** [review]

**Review by allnes — APPROVED** [review]

**Review by AndreySorokin7 — APPROVED** [review]

**Pull request merged**